### PR TITLE
storm-pars, towards new command line interface & API

### PR DESCRIPTION
**Pull request by sjunges**
- [x] First merge https://github.com/moves-rwth/storm/pull/359.
- [x] I have to update the website.  I will make a separate PR for this here: https://github.com/moves-rwth/storm/pull/426
- [x] I have to check stormpy: https://github.com/moves-rwth/stormpy/pull/138

### CODE REVIEW

**Comment by sjunges** [comment]
I still have to push the tests...

**Comment by sjunges** [comment]
@volkm @tquatmann It would be great if you could have a look. Potentially start with the website. Please note that I understand that the current version does not solve all issues we have with the hardly-maintained code base here, but I think this is a step in the right direction.

File: src/storm-pars-cli/print.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+#include "storm/modelchecker/results/CheckResult.h"
+#include "storm/utility/Stopwatch.h"
+
+#include "storm-pars/utility/parametric.h"
+
+namespace storm::pars {
+template<typename ValueType>
+void printInitialStatesResult(std::unique_ptr<storm::modelchecker::CheckResult> const &result,
+                              storm::utility::Stopwatch *watch = nullptr,
+                              const storm::utility::parametric::Valuation<ValueType> *valuation = nullptr);
```
Comment (tquatmann) [inline src/storm-pars-cli/print.h:11]:
```suggestion
                              storm::utility::Stopwatch const* watch = nullptr,
                              const storm::utility::parametric::Valuation<ValueType> *valuation = nullptr);
```
or `std::optional` or `storm::OptionalRef`

File: src/storm-pars/settings/modules/MonotonicitySettings.cpp [hunk 2]
```diff
@@ -13,8 +13,6 @@ namespace storm {
         namespace modules {
     // TODO @Svenja, check what the module prefix is, maybe instead of doing mon- we could set this to true for the onces where we now have mon-"optionname"
```
Comment (tquatmann) [inline src/storm-pars/settings/modules/MonotonicitySettings.cpp:14]:
```suggestion
```

**Review by tquatmann — APPROVED** [review]
Thanks for doing all this! This new code and CLI is much more clean!

While we're at it, I suggest to replace a few more `boost::optional` usages by e.g. `std::optional`, especially in the API code.

File: src/storm-pars-cli/solutionFunctions.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            if (result) {
+                result->filter(storm::modelchecker::SymbolicQualitativeCheckResult<DdType>(model->getReachableStates(), model->getInitialStates()));
+            }
+            return result;
+        },
+        [&model](std::unique_ptr<storm::modelchecker::CheckResult> const& result) {
+            auto parametricSettings = storm::settings::getModule<storm::settings::modules::ParametricSettings>();
+            if (parametricSettings.exportResultToFile() && model->isOfType(storm::models::ModelType::Dtmc)) {
+                auto dtmc = model->template as<storm::models::symbolic::Dtmc<DdType, ValueType>>();
+                // TODO something is broken here
```
[the remark this answers is not in the document]
Comment (sjunges) [reply src/storm-pars-cli/solutionFunctions.h:69]: Thanks for spotting this. It has been fixed (and there is a warning that clarifies one remaining lacking feature)

**Comment by sjunges** [comment]
Many thanks for closely looking at this. I fixed the aspects. I see plenty of possibilities for code improvements, but would like to prioritize other open issues before continueing my efforts towards a more stable parameter code base.

**Review by tquatmann — APPROVED** [review]
Thanks for addressing all my comments. LGTM

**Pull request merged**